src/ray/object_manager/pull_manager.h

```diff
@@ -0,0 +1,264 @@
+  /// is also called when an object appears locally.
+  ///
+  /// \param object_id The ID of the object whose to pull.
+  /// \return Void.
+  void CancelPullObject(const ObjectID &object_id);
+
+  /// Notify the PullManager that a remote object manager wishes to push an
+  /// object chunk to this object manager. Note that this will happen once per
+  /// chunk, not once per object.
+  ///
+  /// \param object_id The ID of the object that is being pushed.
+  /// \param client_id The ID of the remote object manager that is pushing the
+  /// object.
+  /// \param chunk_index The index of the chunk that is being pushed.
+  /// \param num_chunks The total number of chunks that the object is divided
+  /// into.
+  /// \return Void.
+  void ReceivePushRequest(const ObjectID &object_id, const ClientID &client_id,
+                          int64_t chunk_index, int64_t num_chunks);
+
+  /// Notify the PullManager that the locations of the object in the object
+  /// table have changed.
+  ///
+  /// \param object_id The ID of the object whose locations have changed.
+  /// \param clients_with_object The IDs of the object managers that have the
+  /// object.
+  /// \return Void.
+  void NewObjectLocations(const ObjectID &object_id,
+                          const std::unordered_set<ClientID> &clients_with_object);
+
+  /// Notify the PullManager that a chunk was read successfully.
+  ///
+  /// \param object_id The ID of the object that was read.
+  /// \param client_id The ID of the remote object manager that the object was
+  /// read from.
+  /// \param chunk_index The index of the chunk.
+  /// \return Void.
+  void ChunkReadSucceeded(const ObjectID &object_id, const ClientID &client_id,
+                          int64_t chunk_index);
+
+  /// Notify the PullManager that a chunk was not successfully read. This could
+  /// happen because the chunk was intentionally ignored, because the object
+  /// already existed in the store, because the remote object manager died, or
+  /// because the object store was full or the object was already present in the
+  /// object store when we tried to create a chunk.
+  ///
+  /// \param object_id The ID of the object that was read.
+  /// \param client_id The ID of the remote object manager that the object was
+  /// read from.
+  /// \param chunk_index The index of the chunk.
+  /// \return Void.
+  void ChunkReadFailed(const ObjectID &object_id, const ClientID &client_id,
+                       int64_t chunk_index);
+
+  /// Print out a human-readable string describing the PullManager's state.
+  ///
+  /// \return A human-readable string.
+  std::string DebugString() const;
+
+ private:
+  /// Handle the fact that the timer for a pull has expired.
+  ///
+  /// \param object_id The ID of the object that the pull is for.
+  /// \return Void.
+  void TimerExpires(const ObjectID &object_id);
+
+  /// The total number of times PullObject has been called.
```

Comment:
Seems potentially more useful to only keep track of a window of the most recent events for these stats.